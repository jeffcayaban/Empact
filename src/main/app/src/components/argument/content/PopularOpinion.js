import React from 'react';
import { ListGroup, ListGroupItem } from "react-bootstrap";
import { renderArgumentCreatorName } from "../../../utils/Helpers";
import ContentPreviewBtns from "../../common/ContentPreviewBtns";
import ArgumentSources from "../../common/ArgumentSources";

/**
 * Will display the contents of an argument based on popular opinion.
 */

class PopularOpinion extends React.PureComponent {

    render() {
        const { argument, labels, showPetition, showArg, includePreview } = this.props;
        const { explanation, parentArgumentId } = argument;
        const creatorName = renderArgumentCreatorName(argument);
        const argumentTitle = (argument.parentArgumentId !== null && labels[2] !== undefined) ? labels[2] : labels[0];

        const contentPreviewBtns = includePreview && <ContentPreviewBtns parentArgId={parentArgumentId}
                                                                         showArg={showArg} showPetition={showPetition} />;

        return (
            <ListGroup>
                <ListGroupItem>
                    <p className={"popularOpinionSummary"}>
                        {`${creatorName} believes it is generally well known that ${argumentTitle.toLowerCase()}.
                        ${creatorName} has provided the following explanation to prove this:`}
                    </p>
                    <p className={"argumentExplanation"}>{`"${explanation}"`}</p>
                </ListGroupItem>
                <ArgumentSources argument={argument} />
                { contentPreviewBtns }
            </ListGroup>
        );
    }
}

export default PopularOpinion;