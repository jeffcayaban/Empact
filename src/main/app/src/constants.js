export const API_BASE_URL = 'http://localhost:5000/api';
export const ACCESS_TOKEN = 'accessToken';
export const CURRENT_USERNAME = 'username';

export const AUTHORIZATION_HEADER_FIELD = "Authorization";
export const AUTHORIZATION_TOKEN_TYPE = "Bearer";

// Website Name
export const WEBSITE_NAME = "Empact";

// User Roles
export const ROLE_USER = "ROLE_USER";
export const ROLE_ADMIN = "ROLE_ADMIN";

// Argument Types
export const EXPERT_OPINION = "ExpertOpinion";
export const POPULAR_OPINION = "PopularOpinion";

// Argumentation Schemes
export const ARGUMENTATION_SCHEMES = [
    'By Expert Opinion',
    'By Popular Opinion'
];

// Critical Question IDs
export const CRITICAL_QUESTION_IDS = [
    'CQ1',	// Are the circumstances true?
    'CQ2',
    'EO_CQ1',
    'PO_CQ1'
];

// Content Types
export const USER = "User";
export const PETITION = "Petition";
export const ARGUMENT = "Argument";

// Content Properties
export const USERNAME = "Username";
export const ID = "Id";

// Content Actions
export const DELETE = "Delete";
export const CLOSE = "Close";

// Sort Types
export const DESCENDING = "desc";
export const ASCENDING = "asc";
export const SORT_BY_VALUES = [DESCENDING, ASCENDING];

// Content Sort By Values
export const NEWEST = "Newest";
export const OLDEST = "Oldest";
export const MOST_DISCUSSED = "Most Discussed";
export const LEAST_DISCUSSED = "Least Discussed";
export const DISPLAY_SORT_BY_VALUES = [NEWEST, OLDEST, MOST_DISCUSSED, LEAST_DISCUSSED];

// Used in Homepage
export const NO_OF_MOST_DISCUSSED_PETITIONS = 3;
export const NO_OF_MOST_DISCUSSED_ARGUMENTS = 3;

// Petition Form Fields
export const PETITION_TITLE_ID = "title";
export const PETITION_SITUATION_ID = "situation";
export const PETITION_ACTION_ID = "action";
export const PETITION_GOAL_ID = "goal";
export const PETITION_VALUE_ID = "value";
export const PETITION_CLOSING_DATE_ID = "closingDateTime";
export const PETITION_IS_ANONYMOUS_ID = "isAnonymous";

// Register User Fields
export const USERNAME_ID = "username";
export const FIRST_NAME_ID = "firstName";
export const LAST_NAME_ID = "lastName";
export const PASSWORD_ID = "password";
export const CONFIRM_PASSWORD_ID = "confirm-password";

// Register User Field Validation
export const PASSWORD_LENGTH_MIN = 5;
export const PASSWORD_LENGTH_MAX = 21;
export const LAST_NAME_LENGTH_MAX = 40;
export const FIRST_NAME_LENGTH_MAX = 40;
export const USERNAME_LENGTH_MAX = 15;
export const MAXIMUM_NO_OF_SOURCES = 2;
export const MINIMUM_NO_OF_SOURCES = 0;

// Petition Form Validation
export const TITLE_LENGTH_MAX = 90;
export const PETITION_FIELD_MAX = 3000;

// Argument Form Validation
export const EXPERT_FIELD_MAX = 90;
export const ARGUMENT_FIELD_MAX = 6000;

// For Argument Data Visualisation
export const TREE_NETWORK_HEIGHT = 400;

// Other
export const VALID_URL_PREFIX_REGEX = /^(https:\/\/|http:\/\/).*/;
export const PETITION_LIST_SIZE = 5;
export const NOTIFICATION_AUTO_CLOSE_MS = 5000;
export const PETITION_ENTRY_SITUATION_MAX_LENGTH = 400;
export const PETITION_ENTRY_SITUATION_ALT_MAX_LENGTH = 200;
export const MOBILE_MAX_WIDTH = 768;