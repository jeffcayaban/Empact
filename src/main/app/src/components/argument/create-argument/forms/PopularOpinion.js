import React from 'react'
import { createArgumentByPopularOpinion, editArgumentByPopularOpinion } from "../../../../utils/AppUtils";
import { handleArgumentSubmit } from "../../../../utils/Helpers";
import {ARGUMENT_FIELD_MAX, MAXIMUM_NO_OF_SOURCES, MINIMUM_NO_OF_SOURCES} from "../../../../constants";
import ArgumentFormBase from "./argument-form/ArgumentFormBase";
import ArgumentFormField from "./argument-form/ArgumentFormField";

/**
 * Represents the form for an argument by popular opinion.
 */

class PopularOpinion extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            userExplanation: '',
            sources: [],
            isAnonymous: false
        };

        this.createArgRequest = this.createArgRequest.bind(this);
        this.createEditArgRequest = this.createEditArgRequest.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleChangeSource = this.handleChangeSource.bind(this);
        this.handleAddSource = this.handleAddSource.bind(this);
        this.handleRemoveSource = this.handleRemoveSource.bind(this);
        this.handleChangeAnonymity = this.handleChangeAnonymity.bind(this);
        this.onExplanationChange = this.onExplanationChange.bind(this);
    }

    // Used for updating the argument's explanation within the state.
    onExplanationChange(event) {
        const explanation = event.target.value;
        this.setState({ userExplanation: explanation });
    }

    // Used for adding a new source.
    handleAddSource() {
        if (this.state.sources.length <= MAXIMUM_NO_OF_SOURCES) {
            this.setState((prevState) => ({ sources: [...prevState.sources, ""] }));
        }
    }

    // Used for removing an existing source.
    handleRemoveSource() {
        if (this.state.sources.length > MINIMUM_NO_OF_SOURCES) {
            this.setState((prevState) => ({
                sources: prevState.sources.splice(0, prevState.sources.length - 1)
            }));
        }
    }

    // Used for updating an existing source.
    handleChangeSource(index, event) {
        const sources = this.state.sources;
        sources[index] = event.target.value;
        this.setState({ sources: sources });
    }

    // Used for updating an argument's anonymity.
    handleChangeAnonymity() {
        this.setState({ isAnonymous: !this.state.isAnonymous });
    }

    // Create a request object for editing an argument.
    createEditArgRequest() {
        const { isForEdit, argument } = this.props;

        if (isForEdit && argument) {
            const { userExplanation, sources, isAnonymous } = this.state;

            return {
                id: argument.id,
                sources: sources,
                isAnonymous: isAnonymous,
                explanation: userExplanation
            };
        }
    }

    // Create a request object for creating a new argument.
    createArgRequest() {
        if (!this.props.isForEdit) {
            const { isSupporting, petitionId, argumentId, criticalQuestionId, agreesWithCQ, subCriticalQuestionId } = this.props;
            const { userExplanation, sources, isAnonymous } = this.state;

            const request = {
                isSupporting: isSupporting,
                rootPetitionId: petitionId,
                parentArgumentId: argumentId,
                criticalQuestionId: criticalQuestionId,
                agreesWithCQ: agreesWithCQ,
                explanation: userExplanation,
                sources: sources,
                isAnonymous: isAnonymous,
                subCriticalQuestionId: subCriticalQuestionId
            };

            return request;
        }
    }

    // Used for submitting an argument request.
    handleSubmit(e) {
        e.preventDefault();

        const createArgRequest = this.createArgRequest();
        const editArgRequest = this.createEditArgRequest();

        const { isForEdit, argument } = this.props;
        const argumentId = argument ? argument.id : null;

        const sources = isForEdit ? editArgRequest.sources : createArgRequest.sources;

        handleArgumentSubmit(sources, isForEdit, editArgRequest, editArgumentByPopularOpinion, createArgRequest,
            createArgumentByPopularOpinion, argumentId, argument);
    }

    componentWillMount() {
        const argument = this.props.argument;
        if (argument) {
            const { explanation, isAnonymous } = argument;
            const argumentSources = argument.sources && argument.sources.slice();

            this.setState({
                userExplanation: explanation,
                sources: argumentSources,
                isAnonymous: isAnonymous
            });
        }
    }

    render() {
        const statement = this.props.statement;
        const { sources, isAnonymous, userExplanation } = this.state;
        const { handleSubmit, onExplanationChange, handleChangeSource, handleAddSource, handleRemoveSource,
            handleChangeAnonymity  } = this;

        return (
            <form onSubmit={handleSubmit}>
                { statement && <p align="center" style={{ paddingBottom: '20px' }}>{statement}</p> }
                <ArgumentFormField
                    id="formControlsTextarea"
                    label="Enter your explanation in the box below."
                    placeholder="Your explanation"
                    onChange={onExplanationChange}
                    componentClass="textarea"
                    value={userExplanation}
                    className={"petitionTextAreaField"}
                    maxLength={ARGUMENT_FIELD_MAX}
                    required
                />
                <ArgumentFormBase
                    sources={sources}
                    handleChangeSource={handleChangeSource}
                    handleAddSource={handleAddSource}
                    handleRemoveSource={handleRemoveSource}
                    isAnonymous={isAnonymous}
                    handleChangeAnonymity={handleChangeAnonymity}
                />
            </form>
        );
    }
}

export default PopularOpinion;
