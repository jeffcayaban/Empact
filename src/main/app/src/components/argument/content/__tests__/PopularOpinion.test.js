import React from 'react';
import toJson from "enzyme-to-json";
import { shallow } from "enzyme";
import {CRITICAL_QUESTION_IDS, POPULAR_OPINION, ROLE_ADMIN} from "../../../../constants";
import PopularOpinion from "../PopularOpinion";

describe('PopularOpinion', () => {

    it('should render correctly', () => {

        const mockArgument = {
            id: 'testId',
            isSupporting: true,
            rootPetitionId: 'testRootPetitionId',
            parentArgumentId: 'testParentArgumentId',
            subCriticalQuestionId: null,
            agreesWithCQ: true,
            criticalQuestionId: CRITICAL_QUESTION_IDS[0],
            creationDateTime: 1551716401541,
            lastUpdatedDateTime: 1551716402000,
            createdBy: {
                id: 'testId',
                username: 'testUsername',
                firstName: 'testFirstName',
                lastName: 'testLastName',
                authorities: [ROLE_ADMIN]
            },
            argumentType: POPULAR_OPINION,
            isAnonymous: false,
            sources: [],
            explanation: 'testExplanation'
        };

        const component = shallow(<PopularOpinion
            argument={mockArgument}
            labels={[ "test1", "test2", "test3" ]}
            showPetition={jest.fn()}
            includePreview={true}
        />);

        expect(toJson(component)).toMatchSnapshot();
    });
});
