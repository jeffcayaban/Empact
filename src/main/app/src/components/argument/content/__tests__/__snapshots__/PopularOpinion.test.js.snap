// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopularOpinion should render correctly 1`] = `
<ListGroup
  bsClass="list-group"
>
  <ListGroupItem
    bsClass="list-group-item"
    listItem={false}
  >
    <p
      className="popularOpinionSummary"
    >
      testFirstName believes it is generally well known that test3.
                        testFirstName has provided the following explanation to prove this:
    </p>
    <p
      className="argumentExplanation"
    >
      "testExplanation"
    </p>
  </ListGroupItem>
  <ArgumentSources
    argument={
      Object {
        "agreesWithCQ": true,
        "argumentType": "PopularOpinion",
        "createdBy": Object {
          "authorities": Array [
            "ROLE_ADMIN",
          ],
          "firstName": "testFirstName",
          "id": "testId",
          "lastName": "testLastName",
          "username": "testUsername",
        },
        "creationDateTime": 1551716401541,
        "criticalQuestionId": "CQ1",
        "explanation": "testExplanation",
        "id": "testId",
        "isAnonymous": false,
        "isSupporting": true,
        "lastUpdatedDateTime": 1551716402000,
        "parentArgumentId": "testParentArgumentId",
        "rootPetitionId": "testRootPetitionId",
        "sources": Array [],
        "subCriticalQuestionId": null,
      }
    }
  />
  <ContentPreviewBtns
    parentArgId="testParentArgumentId"
    showPetition={[MockFunction]}
  />
</ListGroup>
`;
